### ISSUE TITLE
[java] UnusedPrivateMethod false-positive used in lambda

### ISSUE THREAD

**Issue by jackiemazy** [pmd/pmd#5184]
**Affects PMD Version:**  7.3.0

**Rule:**

Please provide the rule name and a link to the rule documentation:
[https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#unusedprivatemethod](https://docs.pmd-code.org/pmd-doc-7.3.0/pmd_rules_java_bestpractices.html#unusedprivatemethod)

**Description:**
Avoid unused private methods such as 'registerUser(String, String, String)'.

**Code Sample demonstrating the issue:**

```java
public User methodA() {
    val user = userOpt.orElseGet(() -> {
        try {
            return registerUser(email, firstName, lastName);
        } catch (Exception e) {
           throw new IllegalStateException("Failed to register user for " + email, e);
        }
    };
   // ...
   return user;
}

private User registerUser(String email, String firstName, String lastName) throws Exception {
    // register user logic here... 
}
```

**Expected outcome:**

No UnusedPrivateMethod violation should be reported here.

PMD reports a violation at line of the private method 'registerUser' but that's wrong. That's a false positive.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
[Maven] with maven-pmd-plugin version 3.25.

**Comment by jackiemazy** [comment]
@adangel Just tried with pmd 7.5.0 and got the same error:

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.25.0:check (default-cli) on project xxx: PMD 7.5.0 has found 1 violation. For more details see: C:\...\target\pmd.xml -> [Help 1]

**Comment by adangel** [comment]
> Just tried with pmd 7.5.0 and got the same error:

Thanks for double checking!

**Comment by CaptainGlac1er** [comment]
Also had this. another test case would be 

```java
Optional.ofNullable(someVariable)
   .map(this::somePrivateFunction)  // this private function
   .collect(Collectors.toList())
```

**Comment by CaptainGlac1er** [comment]
It also happened with 7.5.0 for me

**Comment by adangel** [comment]
Hi @jackiemazy and @CaptainGlac1er ,

unfortunately I cannot reproduce any of your examples using the following (syntactically correct (!)) code:

```java
package net.sourceforge.pmd.lang.java.rule.bestpractices.unusedprivatemethod;

import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class User {
    private Optional<User> userOpt;

    public User methodA(String email, String firstName, String lastName) {
        User user = userOpt.orElseGet(() -> {
            try {
                return registerUser(email, firstName, lastName);
            } catch (Exception e) {
                throw new IllegalStateException("Failed to register user for " + email, e);
            }
        });
        // ...
        return user;
    }

    private User registerUser(String email, String firstName, String lastName) throws Exception {
        // register user logic here...
        return null;
    }

    public void methodB() {
        User someVariable = new User();
        Optional.ofNullable(someVariable)
                .map(this::somePrivateFunction)  // this private function
                .map(Stream::of)
                .get()
                .collect(Collectors.toList());
    }

    private String somePrivateFunction(User u) {
        return u.toString();
    }
}
```

Correctly identifying method calls inside lambda (including method references) was a problem in the past, see e.g. #4817

Could you provide more details - better even a complete reproducer project?

**Comment by mfvanek** [comment]
I also faced with this issue. Maven. PMD 7.6.0
See https://github.com/EnKarin/BookCrossingServer/pull/228

**Comment by VitaliiIevtushenko** [comment]
The same issue after upgrading from PMD 6.55.0 to 7.7.0
One of the examples where buildItem(a, b) method is considered as unused:
```
        List<SummaryDto.ItemDto> items = new ArrayList<>();
        loads.stream()
                .collect(Collectors.groupingBy(Item::getValue))
                .forEach((a, b) -> items.add(buildItem(a, b)));

        private SummaryDto.ItemDto buildItem(BigDecimal a, List<Item> b) {
            return SummaryDto.ItemDto.builder()....build();
        }
```
In the case of creating the same method without params - there is no issue
```
        List<SummaryDto.ItemDto> items = new ArrayList<>();
        loads.stream()
                .collect(Collectors.groupingBy(Item::getValue))
                .forEach((a, b) -> items.add(buildItem()));

        private SummaryDto.ItemDto buildItem() {
            return SummaryDto.ItemDto.builder()....build();
        }
```

**Comment by lolo101** [comment]
@VitaliiIevtushenko could you provide a complete minimal example?

I've tried to produce a minimal code from your snippet but the issue didn't appear.

**Comment by lolo101** [comment]
I managed to create a working example
Complete Maven project: https://github.com/lolo101/pmd-issue

```java
package org.example.unusedPrivateMethod;

import static java.util.Collections.emptySet;

import java.util.*;
import java.util.stream.*;

public class Main {

    public static void main(String[] args) {
        Library library = new Library(emptySet());
        Map<String, Map<String, String>> map = new Main().run(library);
        System.out.println(map);
    }

    private Map<
        String,
        Map<String, String>
        > run(Library library) {
        return library
            .books()
            .stream()
            .map(
                book ->
                    book
                        .lenders()
                        .stream()
                        .collect(
                            Collectors.toMap(
                                Lender::name,
                                lender ->
                                    Map.of(
                                        book.title(),
                                        lender.status()
                                    )
                            )
                        )
            )
            .reduce(this::reduceBooksAndLenderStatusByLender)
            .orElse(null);
    }

    private Map<
        String,
        Map<String, String>
        > reduceBooksAndLenderStatusByLender(
        Map<String, Map<String, String>> previousMap,
        Map<String, Map<String, String>> nextMap
    ) {
        previousMap.putAll(nextMap);
        return previousMap;
    }
}

record Library(Collection<Book> books) {}
record Book(String title, Collection<Lender> lenders) {}
record Lender(String name, String status) {}
```

**Comment by deripas** [comment]
I also have problem with UnusedPrivateMethod, [example case here](https://github.com/deripas/pmd-example)

```java
package checkstyle;

import java.io.InputStream;

public class BadClass {

    public int Calc(int a, int b) {
        int c = 0;
        Optional.ofNullable("someVariable")
                .map(this::somePrivateFunction)  // this private function false-positive 
                .collect(Collectors.toList());
        return a + b * 31;
    }

    private String somePrivateFunction(String s) {
        return s;
    }
}
```

**Comment by lolo101** [comment]
@deripas the code does not compile : 
```
cannot find symbol
  symbol:   method collect(java.util.stream.Collector<java.lang.Object,capture#1 of ?,java.util.List<java.lang.Object>>)
  location: class java.util.Optional<java.lang.String>
```

**Comment by deripas** [comment]
@lolo101  Yes, you are right, thank you!
Fixed. My problem was in incorrect stream code. After fix stream _UnusedPrivateMethod_ does not triggered.
```java
        Optional.ofNullable("someVariable")
            .map(this::somePrivateFunction) 
            .stream()
            .collect(Collectors.toList());
```

My bad, I apologize for disturbing you)

**Comment by lolo101** [comment]
@deripas no problem 😃
I had a hard time reproducing the issue and my code is quite complicated. I'd like to have a simpler working code.
I didn't mean to be rude, i'm just looking forward a simple working example

**Comment by oowekyala** [comment]
> I managed to create a working example Complete Maven project: https://github.com/lolo101/pmd-issue
> 
> ```java
> package org.example.unusedPrivateMethod;
> 
> import static java.util.Collections.emptySet;
> 
> import java.util.*;
> import java.util.stream.*;
> 
> public class Main {
> 
>     public static void main(String[] args) {
>         Library library = new Library(emptySet());
>         Map<String, Map<String, String>> map = new Main().run(library);
>         System.out.println(map);
>     }
> 
>     private Map<
>         String,
>         Map<String, String>
>         > run(Library library) {
>         return library
>             .books()
>             .stream()
>             .map(
>                 book ->
>                     book
>                         .lenders()
>                         .stream()
>                         .collect(
>                             Collectors.toMap(
>                                 Lender::name,
>                                 lender ->
>                                     Map.of(
>                                         book.title(),
>                                         lender.status()
>                                     )
>                             )
>                         )
>             )
>             .reduce(this::reduceBooksAndLenderStatusByLender)
>             .orElse(null);
>     }
> 
>     private Map<
>         String,
>         Map<String, String>
>         > reduceBooksAndLenderStatusByLender(
>         Map<String, Map<String, String>> previousMap,
>         Map<String, Map<String, String>> nextMap
>     ) {
>         previousMap.putAll(nextMap);
>         return previousMap;
>     }
> }
> 
> record Library(Collection<Book> books) {}
> record Book(String title, Collection<Lender> lenders) {}
> record Lender(String name, String status) {}
> ```

I've been looking at this but it is difficult to debug. At least I can reproduce it. I'm not sure it is an instance of the original issue though. Will update later

**Comment by deripas** [comment]
@lolo101   How does code compilation work? I encountered such an interesting situation: try check file [BadClass.java](https://github.com/deripas/pmd-example/blob/main/src/test/resources/BadClass.java)

```java
package com.semrush.siteaudit.crowly.kafka.component;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.concurrent.CompletableFuture;

/**
 * Test.
 */
@ExtendWith(MockitoExtension.class)
class SomeTest {

    @Test
    void test() {
        final BytesListener listener = createListener(
            (bytes) -> CompletableFuture.completedFuture("HI!")
        );
        Assertions.assertNotNull(listener.onRecord(new byte[0]));
    }

    private static BytesListener createListener(
        BytesParser<String> parser
    ) {
        return bytes -> parser
            .parse(bytes)
            .thenAccept(System.out::println);
    }
}

```
OK - no issues, if there is a class [BytesParser.java](https://github.com/deripas/pmd-example/blob/main/src/test/java/com/github/deripas/pmd/BytesParser.java) in ClassPath.

As soon as I delete the class [BytesParser.java](https://github.com/deripas/pmd-example/blob/main/src/test/java/com/github/deripas/pmd/BytesParser.java)  (argument type) that`UnusedPrivateMethod` is triggered.
```
Avoid unused private methods such as 'createListener(BytesParser<String>)'
```
Removing the second class  [BytesListener.java](https://github.com/deripas/pmd-example/blob/main/src/test/java/com/github/deripas/pmd/BytesListener.java) (method return type) does not trigger the issue t`UnusedPrivateMethod`.

Why is it important for a static analyzer to have a neighbor class?

**Comment by jsotuyod** [comment]
@lolo101 to understand if a method is used or not, we need to find usages. Finding usages is a multi-step process that on a **very** high level goes like this:

1. Find "candidate" methods (symbols that match by name)
2. Check applicability of each of those methods (argument number + types)
3. Sort out the priority (which apply more closely), ie: does it require unboxing? type coercion?

This second step is particularly complex (varargs, boxing / unboxing, lambdas, generics, implemented interfaces, etc.)

Without argument type data, we can't tell if a method is a match or not based on symbol names alone.

Knowing the shape of `BytesParser<String>` means, knowing the class, understanding it's a functional interface, that the only method, is a `(bytes) -> CompletableFuture<T>` where `T` in this instance is `String`, so we know `(bytes) -> CompletableFuture.completedFuture("HI!")` is a match, and the method is actually invoked!

Java doesn't consider the return types as part of the matching algorithm (think about, overriding a method with same args but different return type is a compilation error). So for this particular case, we *need*  `BytesParser`,  but we don't need `BytesListener`.

PMD 7.0.0 actually implements the Java Language Specification on this, which greatly improved our accuracy vs PMD 6; but is also among the most complex parts of PMD as it requires understanding the JLS in detail.

Also of note, we actually need the `. class`, not the `.java` file, as the way we do this works regardless of wether the types are defined within the codebase, or in a dependency.

For this particular scenario, where there are no static wildcard imports, we are not extending any other class / implementing any interface and there is no override for the invoked method we *could* assume it's a match, but to be honest, there are so many constraints it doesn't make sense. Moreover, type data is used in most rules, and we can't always make these assumptions.

PMD gives for granted code is valid, and once bound into the build pipeline (ie: Maven / Gradle) getting the needed data is pretty much free.

**Comment by oowekyala** [comment]
> At the end of the day, the quality of the outputs are a function of the quality of the inputs.

Totally agree. The easiest way to fix this issue, and to improve PMD results for many rules, is to configure your classpath properly. There is only so much we can do without proper configuration.

That being said, our type resolution code is pretty error-tolerant. Even if you don't have a full classpath, we don't give up on type resolution that easily. For instance in the example in the original post:
```java
class Foo {
public User methodA() {
    val user = userOpt.orElseGet(() -> {
        try {
            return registerUser(email, firstName, lastName);
        } catch (Exception e) {
           throw new IllegalStateException("Failed to register user for " + email, e);
        }
    });
   // ...
   return user;
}

private User registerUser(String email, String firstName, String lastName) throws Exception {
    // register user logic here... 
}
}
```
I adapted it to be syntactically valid, but it is still semantically invalid as it refers to undefined variables (like `email`, `firstName`, etc). "Undefined" here is just the same as "defined somewhere else, but that place is not on the classpath", PMD cannot know the difference.

In the code above, we still do type inference, but give the undefined variables the special type `(*unknown*)`, which is compatible with any type. So the method call is indeed resolved to the method registerUser, and it should not be marked unused. So I cannot reproduce the example in the original post in any case.

In the call chain in https://github.com/pmd/pmd/issues/5184#issuecomment-2460116799, there is indeed a reproducible bug.
```java
  List<SummaryDto.ItemDto> items = new ArrayList<>();
        loads.stream()
                .collect(Collectors.groupingBy(Item::getValue))
                .forEach((a, b) -> items.add(buildItem(a, b)));

        private SummaryDto.ItemDto buildItem(BigDecimal a, List<Item> b) {
            return SummaryDto.ItemDto.builder()....build();
        }
```
If your classpath is invalid, (say `Item` is absent, but `loads` is present and is a `List<Item>`, and all `java.util` types are on the classpath and properly imported), then the `Item::getValue` returns an unknown type. Nevertheless type resolution _should_ infer that `collect` has type `Map<(*unknown*), List<Item>>`. Then the parameters `a` and `b` of the lambda have types `(*unknown*)` and `List<Item>` respectively, and `buildItem` is correctly resolved to the private method.

Here the bug is that `collect` is inferred instead to `Map<Object, List<Item>>`, which doesn't behave the same as it will cause the overload resolution for `buildItem` to fail, so that `buildItem` in the end is marked as unused. As was shown in the comment, `buildItem()` without parameters is correctly resolved as it does not depend on the inferred types of `a` or `b`. I extracted that issue in #5329 and fixed it in #5330.

**Comment by oowekyala** [comment]
I created #5338 for this comment: https://github.com/pmd/pmd/issues/5184#issuecomment-2471680894

I am closing this as the original issue is not reproducible. The other issues that were reported in this thread have been extracted into their own tickets and fixed.
